### PR TITLE
Use non-BOM encodings

### LINKED ISSUE #2369: Possible bug in reading zero width no-break space character
### Environment

-   Pythonnet version: 3.0.3
-   Python version: 3.10
-   Operating System: Windows 11
-   .NET Runtime: 8.0.4

### Details

-   Describe what you were trying to get done.

I want to read a file that may contain zero width no-break space characters anywhere in the file. If I read a file using C# into Python using Python.NET, these characters get dropped if they are at the start of a line.

-   What commands did you run to trigger this issue? If you can provide a
    [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
    this will help us understand the issue.

```python
from pythonnet import load
load("coreclr")
import clr
clr.AddReference('System')
clr.AddReference('System.Runtime')
clr.AddReference('System.IO')

import System
file_path = r"path\to\zwnbsp_test_data.txt"

result = System.IO.File.ReadAllLines(file_path)
print(str(result[1][0]))
```

Native C# code does not seem to have this problem:
```
C#
String[] lines = File.ReadAllLines(@"path\to\zwnbsp_test_data.txt");

int firstChar = char.ConvertToUtf32(lines[1], 0);
int secondChar = char.ConvertToUtf32(lines[1], 1);
Console.WriteLine(firstChar);
Console.WriteLine(secondChar);
```

[zwnbsp_test_data.txt](https://github.com/pythonnet/pythonnet/files/15194162/zwnbsp_test_data.txt)

Some screenshots to help illustrate: 
image
image

-   If there was a crash, please include the traceback here.
No crash

### PR DESCRIPTION
**Pull request by filmor**
Use non-BOM encodings for both C#->Python and Python->C#, as the byteorder is always the native one and the BOM is neither never or always used.

Fixes #2369.

### CODE REVIEW

**Comment by lostmsu** [comment]
@filmor can you ELI5? For someone not familiar with intricacies of BOM, but aware of byte order issues.

My biggest question is if this change has any potential to introduce bugs to handling strings that actually have BOM? E.g. imagine a scenario when someone serialized and persisted something with BOM using 3.0.3, but after this change in 3.0.4 if they read it back BOM will be in their string data.

**Comment by filmor** [comment]
It's the other way round. Strings that are being passed between Python and .NET are UTF16 in the respective native byte order (usually LE), without a BOM. The functions that we were using for the conversions (in particular [`PyUnicode_DecodeUTF16`](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_DecodeUTF16) and the defaulr encoding objects from `Encoding`) try to be "smart" and will interpret a leading set of FE FF or FF FE as the byte order mark, removing it from the converted string. By passing the correct endian-ness explicitly, this behaviour is disabled.
